Sources/GRPCCore/Server.swift

```diff
@@ -0,0 +1,428 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Atomics
+
+/// A gRPC server.
+///
+/// The server accepts connections from clients and listens on each connection for new streams
+/// which are initiated by the client. Each stream maps to a single RPC. The server routes accepted
+/// streams to a service to handle the RPC or rejects them with an appropriate error if no service
+/// can handle the RPC.
+///
+/// A ``Server`` may listen with multiple transports (for example, HTTP/2 and in-process) and route
+/// requests from each transport to the same service instance. You can also use "interceptors",
+/// to implement cross-cutting logic which apply to all accepted RPCs. Example uses of interceptors
+/// include request filtering, authentication, and logging. Once requests have been intercepted
+/// they are passed to a handler which in turn returns a response to send back to the client.
+///
+/// ## Creating and configuring a server
+///
+/// The following example demonstrates how to create and configure a server.
+///
+/// ```swift
+/// var server = Server()
+///
+/// // Create and add an in-process transport.
+/// let inProcessTransport = InProcessServerTransport()
+/// server.transports.add(inProcessTransport)
+///
+/// // Create and register the 'Greeter' and 'Echo' services.
+/// server.services.register(GreeterService())
+/// server.services.register(EchoService())
+///
+/// // Create and add some interceptors.
+/// server.interceptors.add(StatsRecordingServerInterceptors())
+/// ```
+///
+/// ## Starting and stopping the server
+///
+/// Once you have configured the server call ``run()`` to start it. Calling ``run()`` starts each
+/// of the servers transports. A ``ServerError`` is thrown if any of the transports can't be
+/// started.
+///
+/// ```swift
+/// // Start running the server.
+/// try await server.run()
+/// ```
+///
+/// The ``run()`` method won't return until the server has finished handling all requests. You can
+/// signal to the server that it should stop accepting new requests by calling ``stopListening()``.
+/// This allows the server to drain existing requests gracefully. To stop the server more abruptly
+/// you can cancel the task running your server. If your services require additional resources that
+/// need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct Server: Sendable {
+  typealias Stream = RPCStream<ServerTransport.Inbound, ServerTransport.Outbound>
+
+  /// A collection of ``ServerTransport`` implementations that the server uses to listen
+  /// for new requests.
+  public var transports: Transports
+
+  /// The services registered which the server is serving.
+  public var services: Services
```

Comment:
I am a bit unsure about making those vars. In the `ServiceGroup` I had the same decision and went with expecting everything in the inits. The reason for this was that you really expect everything to be setup before `run()` is called and allowing users to change them after `run()` is called is a bit weird.

Reply:
IMO these make the API easier to use, passing everything in `init` is quite awkward when there are potentially many things to pass in and is awkward w.r.t. API evolution. Sure, you can wrap everything up in a config, but that just adds more indirection for users.

I don't see changing things after `run()` to be an issue to be honest, trying to make changes after `run()` has been called just makes no sense at all so it's highly unlikely that someone would fall foul to this. We also document quite clearly that this isn't supported.

Reply:
Hm, I am not fully convinced yet. Let's imagine the following

```
var server = Server()
server.transports = [NIOH2Transport(port: 1234)]
async let _ = try await server.run()
server.transports = [NIOH2Transport(port: 5677])
async let _ try await server.run()
```

There really isn't a reason the second run should fail but it does because of the shared state in the struct. At least personally, I find these semantics confusing. I do get the reason why we want a struct here but then maybe we shouldn't share the state across `run` invocations?